x-pack/modules/azure/lib/azure_module_config_generator.rb

```diff
@@ -0,0 +1,104 @@
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+
+require 'logstash/namespace'
+
+module LogStash
+  module Azure
+    class ConfigGenerator
+
+      def generate_input(module_hash)
+        lines = []
+        config = []
+        module_hash.each do |k, v|
+          config << {parse_key(k) => v} if k.start_with?('var.input.azure_event_hubs') && !k.eql?('var.input.azure_event_hubs.event_hubs')
+        end
+        mode = :basic
+        event_hubs = module_hash['var.input.azure_event_hubs.event_hubs']
+        if event_hubs
+          if event_hubs.is_a?(Array) && event_hubs.size > 1 && event_hubs[0][0].eql?('name')
+            reference = event_hubs[0]
+            if reference.is_a?(Array)
+              mode = :advanced
+              # convert parallel arrays to array of hashes
+              advanced_config = []
+              event_hubs.size.times do |x| # process event hubs north -> south
+                hub_array = []
+                next if x == 0 # skip reference array, reference array must be in position 0
+                reference.each_with_index do |key, y| # process the reference array west -> east
+                  next if y == 0 # skip the 'name' in the reference array
+                  hub_array << {key => event_hubs[x][y]} # for each item in the reference array, create a hash of reference key to the correct x,y axis, add to array
+                end
+                advanced_config << {event_hubs[x][0] => hub_array} # for each event hub create a hash keyed by the hub name. Note the hub name must always be position 0.
+              end
+              config << {'event_hubs' => advanced_config} # add the advanced config to the main config... advanced config will always take precedence via the plugin.
+            end
+          else
+            raise "Invalid configuration. 'var.input.azure_event_hubs.event_hubs' must be a set of arrays. Please check your configuration."
+          end
+        end
+
+        basic_valid = false
+        advanced_valid = false
+        lines << '# start azure_event_hubs' # helps with testing
+        # write the config to the buffer
+        config.each do |c|
+          c.each do |k, v|
+            if v.is_a?(Array) && k.eql?('event_hubs') && mode.eql?(:advanced)
```

Comment:
this whole generator is a bit scary with all the string manipulation. I think we could improve this by doing more at the ERB template level, filling values through iterating the arrays in the template, instead of generating 1 big string and injecting it there. thoughts?

Reply:
I intentionally stayed away from doing complex things in the ERB. IMO templates are best suited for simple substitutions without much (if any) business logic. 

The manipulation here is all with Ruby hash/array, the conversion from Ruby hash/array to Logstash config language (the string) is a necessary evil and has hidden complexity no matter where it gets done.